src/chain/store/index.rs

```diff
@@ -151,139 +62,186 @@ impl<DB: Blockstore> ChainIndex<DB> {
-        Ok(lbe)
+    fn persist_tipset(tipset: &Tipset, db: &impl Blockstore) {
+        for block in tipset.blocks() {
+            db.put_cbor_default(block).unwrap();
+        }
     }
 
-    /// Rounds height epoch to nearest sparse cache index epoch.
-    fn round_height(&self, height: ChainEpoch) -> ChainEpoch {
-        (height / SKIP_LENGTH) * SKIP_LENGTH
+    fn genesis_tipset() -> Tipset {
+        Tipset::from(BlockHeader::default())
     }
 
-    /// Gets the closest rounded sparse index and returns the loaded tipset at
-    /// that index.
-    fn round_down(&self, ts: Arc<Tipset>) -> Result<Arc<Tipset>, Error> {
-        let target = self.round_height(ts.epoch());
-
-        self.walk_back(ts, target)
+    fn tipset_child(parent: &Tipset, epoch: ChainEpoch) -> Tipset {
+        // Use a static counter to give all tipsets a unique timestamp
+        static COUNTER: AtomicU64 = AtomicU64::new(0);
+        let n = COUNTER.fetch_add(1, Ordering::Relaxed);
+        Tipset::from(
+            BlockHeader::builder()
+                .parents(parent.key().clone())
+                .epoch(epoch)
+                .timestamp(n)
+                .build()
+                .unwrap(),
+        )
     }
 
-    /// Load parent tipsets until the `to` [`ChainEpoch`].
-    fn walk_back(&self, from: Arc<Tipset>, to: ChainEpoch) -> Result<Arc<Tipset>, Error> {
-        if to > from.epoch() {
-            return Err(Error::Other(
-                "Looking for tipset with height greater than start point".to_string(),
-            ));
-        }
-
-        if to == from.epoch() {
-            return Ok(from);
-        }
+    #[test]
+    fn get_null_tipset() {
+        let db = Arc::new(MemoryDB::default());
+        let gen = genesis_tipset();
+        let epoch1 = tipset_child(&gen, 1);
+        let epoch3 = tipset_child(&epoch1, 3);
+        let epoch4 = tipset_child(&epoch3, 4);
+        persist_tipset(&gen, &db);
+        persist_tipset(&epoch1, &db);
+        persist_tipset(&epoch3, &db);
+        persist_tipset(&epoch4, &db);
+
+        let index = ChainIndex::new(db);
+        // epoch 2 is null. ResolveNullTipset decided whether to return epoch 1 or epoch 3
+        assert_eq!(
+            index
+                .tipset_by_height(2, Arc::new(epoch4.clone()), ResolveNullTipset::TakeOlder)
+                .unwrap()
+                .as_ref(),
+            &epoch1
+        );
 
-        let mut ts = from;
-        loop {
-            let pts = self.load_tipset(ts.parents())?;
+        assert_eq!(
+            index
+                .tipset_by_height(2, Arc::new(epoch4), ResolveNullTipset::TakeYounger)
+                .unwrap()
+                .as_ref(),
+            &epoch3
+        );
+    }
 
-            if to > pts.epoch() {
-                // Pts is lower than to epoch, return the tipset above that height
-                return Ok(ts);
-            }
+    #[test]
+    fn get_different_branches() {
```

Comment:
new issue:

wibni declarative API for building tipsets. I can implement or mentor on this, I've done similar stuff before, and it's _super important_ to make testing easy.
```rust
tipsets!(
    genesis --> epoch1a(timestamp = 1) --> epoch2a,
    genesis --> epoch1b --> epoch2b
)
```

Reply:
#3228